backend/apps/ecommerce/models.py

```diff
@@ -0,0 +1,115 @@
+import uuid
+
+from django.conf import settings
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models, transaction
+from django.db.models import F
+from django.db.models.fields import DateTimeField
+
+from apps.organizations.models import Organization
+
+from ..users.models import User
+
+
+class Product(models.Model):
+    name = models.CharField(max_length=200)
+    price = models.DecimalField(max_digits=11, decimal_places=2)
+    description = models.TextField()
+    organization = models.ForeignKey(Organization, on_delete=models.CASCADE, related_name="products")
+    total_quantity = models.IntegerField()
+    current_quantity = models.IntegerField(null=True)  # Set to total_quantity upon initialization
+    max_buyable_quantity = models.IntegerField(default=1)
+
+    # Generic foreign key to related product model instance (e.g event model)
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, null=True)
+    object_id = models.PositiveIntegerField(null=True)
+    related_object = GenericForeignKey("content_type", "object_id")
+
+    def __str__(self):
+        return self.name
+
+    def save(self, *args, **kwargs):
+        if not self.current_quantity:
+            self.current_quantity = self.total_quantity
+        self.max_buyable_quantity = min(self.max_buyable_quantity, self.total_quantity)
+        super().save(*args, **kwargs)
+
+    @classmethod
+    def check_and_reserve_quantity(cls, product_id, user: User, quantity: int):
+        with transaction.atomic():
+            # Check if the requested quantity is allowed
+            try:
+                # Acquire DB lock for the product (no other process can change it)
+                product = cls.objects.select_for_update().get(pk=product_id)
+            except cls.DoesNotExist:
+                raise ValueError("Ugyldig produkt")
+
+            if product.related_object and not product.related_object.is_user_allowed_to_buy_product(user):
+                raise Exception("Du kan ikke kjøpe dette produktet.")
+
+            captured_orders = Order.objects.filter(
+                product__id=product_id,
+                user=user,
+                payment_status=Order.PaymentStatus.CAPTURED,
+            )
+            bought_quantity = sum([order.quantity for order in captured_orders])
```

Comment:
This can be achieved on DB-level with [aggregate](https://docs.djangoproject.com/en/dev/topics/db/aggregation/#generating-aggregates-over-a-queryset).
```suggestion
            from django.db.models import Sum
            
            bought_quantity = Order.objects
                .filter(
                    product__id=product_id,
                    user=user,
                    payment_status=Order.PaymentStatus.CAPTURED,
                )
                .aggregate(bought_quantity=Sum("quantity"))["bought_quantity"]
```